sympy/combinatorics/grouphomomorphism.py

```diff
@@ -0,0 +1,215 @@
+from __future__ import print_function, division
+
+from sympy.combinatorics.fp_groups import FpGroup, FpSubgroup
+from sympy.combinatorics.free_groups import FreeGroupElement
+from sympy.combinatorics.perm_groups import PermutationGroup
+
+class GroupHomomorphism(object):
+    '''
+    A class representing group homomorphisms. Instantiate using `homomorphism()`.
+
+    References
+    ==========
+    [1] Holt, D., Eick, B. and O'Brien, E. (2005). Handbook of computational group theory.
+
+    '''
+
+    def __init__(self, domain, codomain, images):
+        self.domain = domain
+        self.codomain = codomain
+        self.images = images
+        self._inverses = None
+        self._kernel = None
+        self._image = None
+
+    def _invs(self):
+        '''
+        Return a dictionary with `{gen: inverse}` where `gen` is a rewriting
+        generator of `codomain` (e.g. strong generator for permutation groups)
+        and `inverse` is an element of its preimage
+
+        '''
+        image = self.image()
+        vals = [v for v in self.images.values() if not v.is_identity]
+        keys = list(self.images.keys())
+        inverses = {}
+        for i, v in enumerate(vals):
+            inverses[v] = keys[i]
```

Comment:
It looks like this could be wrong, if `vals` is shorter than `keys`. `keys[i]` could be mapped to identity.